modules/renter/renter.go

```diff
@@ -286,31 +291,66 @@ func (r *Renter) PriceEstimation() modules.RenterPriceEstimation {
 	return est
 }
 
+// setBandwidthLimits will change the bandwidth limits of the renter based on
+// the persist values for the bandwidth.
+func (r *Renter) setBandwidthLimits(downloadSpeed int64, uploadSpeed int64) error {
+	// Input validation.
+	if downloadSpeed < 0 || uploadSpeed < 0 {
+		return errors.New("download/upload rate limit can't be below 0")
+	}
+
+	// Check for sentinal "no limits" value.
+	if downloadSpeed == 0 && uploadSpeed == 0 {
+		r.hostContractor.SetRateLimits(0, 0, 0)
+	} else {
+		// Set the rate limits according to the provided values.
+		r.hostContractor.SetRateLimits(downloadSpeed, uploadSpeed, 4*4096)
+	}
+	return nil
+}
+
 // SetSettings will update the settings for the renter.
+//
+// NOTE: This function can't be atomic. Typically we try to have user requests
+// be atomic, so that either everything changes or nothing changes, but since
+// these changes happen progressively, it's possible for some of the settings
+// (like the allowance) to succeed, but then if the bandwidth limits for example
+// are bad, then the allowance will update but the bandwidth will not update.
 func (r *Renter) SetSettings(s modules.RenterSettings) error {
+	// Early input valudation.
+	if s.MaxDownloadSpeed < 0 || s.MaxUploadSpeed < 0 {
```

Comment:
Is that redundant on purpose? You also check for it in `setBandwidthLimits`.

Reply:
yes, want to catch the issues earlier rather than later